### PR TITLE
test: add unit tests for wallets (funding sources)

### PR DESCRIPTION
**Pull request by motorina0**
All funding sources must behave uniformly. If, for example, the endpoint returns an `HTTP 404` then all wallets api should return a `Pending` state. It is nok for some wallets to return a status and for others to raise an exception.

This PR creates the basis for checking the interface of all rest wallets. It currently only handles `corelightingrest` and `lndrest`, but will easily be extended.

The test data is specified in a JSON fixture. Adding new funding sources means updating the JSON.

A new dev dependency to `pytest-httpserver` has been added. This lib is used to simulate the LN Node rest responses.

Minor changes have been made to `corelightningrest.py`, `lndrest.py` and `base.py` in order to unify the interface.

This PR also closes: https://github.com/lnbits/lnbits/pull/2062

### CODE REVIEW

File: tests/wallets/test_corelightningrest.py [hunk 1]
```diff
@@ -0,0 +1,816 @@
...
+    + "sgk44qyhalv6ust54x98whl3q635hkwgsyw8xgqjl7jwu",
+)
+
+
+# specify where the server should bind to
+@pytest.fixture(scope="session")
+def httpserver_listen_address():
+    return ("127.0.0.1", 8555)
+
+
```
Comment (dni) [inline tests/wallets/test_corelightningrest.py:43]:
```suggestion
@pytest.fixture(scope="module", autouse=True)
def cln_fixture(httpserver: HTTPServer):
       settings.corelightning_rest_url = ENDPOINT
    settings.corelightning_rest_macaroon = MACAROON
    yield param
    httpserver.check_assertions()
```

File: tests/wallets/test_corelightningrest.py [hunk 2]
```diff
@@ -0,0 +1,816 @@
...
+    + "djyukhys0z2rw6grgm8me9k4w9vn05zt"
+    + "9svzzjspp5ud2jdfpaqn5c2k2vphatsj"
+    + "ypfafyk8rcvkvwexnrhmwm94ex4jtqdq"
+    + "u24hxjapq23jhxapqf9h8vmmfvdjscqp"
+    + "jrzjqta942048v7qxh5x7pxwplhmtwfl"
+    + "0f25cq23jh87rhx7lgrwwvv86r90guqq"
+    + "nwgqqqqqqqqqqqqqqpsqyg9qxpqysgqy"
+    + "lngsyg960lltngzy90e8n22v4j2hvjs4"
+    + "l4ttuy79qqefjv8q87q9ft7uhwdjakvn"
+    + "sgk44qyhalv6ust54x98whl3q635hkwgsyw8xgqjl7jwu",
```
Comment (dni) [inline tests/wallets/test_corelightningrest.py:34]:
```suggestion
    "djyukhys0z2rw6grgm8me9k4w9vn05zt"
    "9svzzjspp5ud2jdfpaqn5c2k2vphatsj"
    "ypfafyk8rcvkvwexnrhmwm94ex4jtqdq"
    "u24hxjapq23jhxapqf9h8vmmfvdjscqp"
    "jrzjqta942048v7qxh5x7pxwplhmtwfl"
    "0f25cq23jh87rhx7lgrwwvv86r90guqq"
    "nwgqqqqqqqqqqqqqqpsqyg9qxpqysgqy"
    "lngsyg960lltngzy90e8n22v4j2hvjs4"
    "l4ttuy79qqefjv8q87q9ft7uhwdjakvn"
    "sgk44qyhalv6ust54x98whl3q635hkwgsyw8xgqjl7jwu",
```

javascript :D

File: tests/wallets/test_lndrest.py [hunk 3]
```diff
@@ -0,0 +1,701 @@
+import json
+
+import pytest
+from httpx import HTTPStatusError
+from pytest_httpserver import HTTPServer
+from werkzeug.wrappers import Response
+
+from lnbits.wallets.corelightningrest import settings
```
Comment (dni) [inline tests/wallets/test_lndrest.py:8]: importing settings from `lnbits.wallets.corelightningrest`?

Comment (motorina0) [reply tests/wallets/test_lndrest.py:8]: yes, if I use `from lnbits.settings import settings` then the test and the wallet will have two different instances of `settings`

**Comment by bitkarrot** [comment]
if one of you guys has a spare moment can you please check the unit test here: https://github.com/lnbits/lnbits/pull/2270

thanks in advance. :)

File: lnbits/wallets/lndrest.py [hunk 4]
```diff
@@ -154,29 +178,53 @@ async def pay_invoice(self, bolt11: str, fee_limit_msat: int) -> PaymentResponse
...
-        fee_msat = int(data["payment_route"]["total_fees_msat"])
-        preimage = base64.b64decode(data["payment_preimage"]).hex()
-        return PaymentResponse(True, checking_id, fee_msat, preimage, None)
+            checking_id = base64.b64decode(data["payment_hash"]).hex()
+            fee_msat = int(data["payment_route"]["total_fees_msat"])
+            preimage = base64.b64decode(data["payment_preimage"]).hex()
+            return PaymentResponse(True, checking_id, fee_msat, preimage, None)
+        except json.JSONDecodeError:
+            return PaymentResponse(
+                False, None, None, None, "Server error: 'invalid json response'"
```
Comment (dni) [inline lnbits/wallets/lndrest.py:211]: can we use kwargs in places like that? aka `return PaytmentRepsonse(ok=False, error_msg="invalid...")`

File: lnbits/wallets/base.py [hunk 5]
```diff
@@ -18,6 +18,18 @@ class InvoiceResponse(NamedTuple):
...
+        return self.ok is True
+
+    @property
+    def pending(self) -> bool:
+        return self.ok is None
+
+    @property
+    def failed(self) -> bool:
+        return self.ok is False
+
```
Comment (dni) [inline lnbits/wallets/base.py:32]: hmm. just want to say it here i am not a huge fan of this magic properties, and also that `ok` has 3 states, True, None, False
ignore this comment

File: lnbits/wallets/lndrest.py (hunk 4 above)
Comment (motorina0) [reply lnbits/wallets/lndrest.py:211]: Yes, we can and we should, there is a `PaymentResponse[Success | Failed | Pending]` PR.

In this PR I have made minimum changes to the code, the focus is to introduce the tests so future changes are less risky.

**Review by dni — APPROVED** [review]

**Comment by prusnak** [comment]
Needs rebase to remove conflicts in `poetry.lock` and `pyproject.toml`

**Comment by prusnak** [comment]
CI mypy/pyright failures:

```
tests/wallets/test_rest_wallets.py:7: error: Cannot find implementation or library stub for module named "pytest_httpserver"  [import-not-found]
tests/wallets/test_rest_wallets.py:8: error: Cannot find implementation or library stub for module named "werkzeug.wrappers"  [import-not-found]
tests/wallets/test_rest_wallets.py:8: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

**Comment by motorina0** [comment]
> CI mypy/pyright failures:
> 
> ```
> tests/wallets/test_rest_wallets.py:7: error: Cannot find implementation or library stub for module named "pytest_httpserver"  [import-not-found]
> tests/wallets/test_rest_wallets.py:8: error: Cannot find implementation or library stub for module named "werkzeug.wrappers"  [import-not-found]
> tests/wallets/test_rest_wallets.py:8: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
> ```

Yeah, I've pushed the rebased branch then ran again `poetry add --dev pytest_httpserver`. Easier this way.

**Review by talvasconcelos — APPROVED** [review]

**Pull request merged**